### PR TITLE
[RLlib]: Fix SampleBatch.split_by_episode to use dones if episode id is not available

### PR DESCRIPTION
**Pull request by Rohan138**
## Checks

- [Y] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/policy/sample_batch.py [hunk 1]
```diff
@@ -352,36 +352,34 @@ def split_by_episode(self) -> List["SampleBatch"]:
...
 
-        # No eps_id in data -> Make sure there are no "dones" in the middle
-        # and add eps_id automatically.
-        if SampleBatch.EPS_ID not in self:
-            # TODO: (sven) Shouldn't we rather split by DONEs then and not
-            #  add fake eps-ids (0s) at all?
-            if SampleBatch.DONES in self:
-                assert not any(self[SampleBatch.DONES][:-1])
-            self[SampleBatch.EPS_ID] = np.repeat(0, self.count)
+        slices = []
```
Comment (sven1977) [inline rllib/policy/sample_batch.py:355]: Could we add more comments here as to why we would want to do this?

Also, can we add a test to `test_sample_batch.py` to check this?

File: rllib/policy/sample_batch.py [hunk 2]
```diff
@@ -352,36 +352,34 @@ def split_by_episode(self) -> List["SampleBatch"]:
...
+        if SampleBatch.EPS_ID in self:
+            # Produce a new slice whenever we find a new episode ID.
+            cur_eps_id = self[SampleBatch.EPS_ID][0]
+            offset = 0
+            for i in range(self.count):
+                next_eps_id = self[SampleBatch.EPS_ID][i]
```
Comment (sven1977) [inline rllib/policy/sample_batch.py:361]: Wait, if cur_eps_id points to idx 0 and we start the for-loop with i=0, then I'm always comparing the first element to itself. Should we start the for-loop at idx=1, then?

File: rllib/policy/sample_batch.py [hunk 3]
```diff
@@ -352,36 +352,34 @@ def split_by_episode(self) -> List["SampleBatch"]:
...
+                if next_eps_id != cur_eps_id:
+                    slices.append(self[offset:i])
+                    offset = i
+                    cur_eps_id = next_eps_id
+            # Add final slice.
+            slices.append(self[offset : self.count])
+
+        elif SampleBatch.DONES in self:
```
Comment (sven1977) [inline rllib/policy/sample_batch.py:369]: Also here, please add a comment on top of the `elif`.

File: rllib/policy/sample_batch.py (hunk 2 above)
Comment (Rohan138) [reply rllib/policy/sample_batch.py:361]: No, we still have to append the first episode to `slices`, which we do inside the loop

File: rllib/policy/sample_batch.py [hunk 4]
```diff
@@ -346,42 +347,47 @@ def split_by_episode(self) -> List["SampleBatch"]:
 
         Examples:
             >>> from ray.rllib.policy.sample_batch import SampleBatch
+            >>> # "eps_id" is present
             >>> batch = SampleBatch( # doctest: +SKIP
             ...     {"a": [1, 2, 3], "eps_id": [0, 0, 1]})
             >>> print(batch.split_by_episode()) # doctest: +SKIP
             [{"a": [1, 2], "eps_id": [0, 0]}, {"a": [3], "eps_id": [1]}]
+            >>>
+            >>> # "eps_id" not present, split by "dones" instead
```
Comment (kouroshHakha) [inline rllib/policy/sample_batch.py:356]: will you error out if done is not True at all or for the last episode? What's the behavior then?

File: rllib/policy/sample_batch.py [hunk 5]
```diff
@@ -346,42 +347,47 @@ def split_by_episode(self) -> List["SampleBatch"]:
...
+        # No eps_id in data -> split by dones instead
+        elif SampleBatch.DONES in self:
+            offset = 0
+            for i in range(self.count):
+                if self[SampleBatch.DONES][i]:
+                    slices.append(self[offset : i + 1])
+                    offset = i + 1
+            # Add final slice.
+            if offset != self.count:
+                slices.append(self[offset : self.count])
```
Comment (kouroshHakha) [inline rllib/policy/sample_batch.py:386]: Can you make it `slices.append(self[count:])`?

File: rllib/policy/sample_batch.py [hunk 6]
```diff
@@ -346,42 +347,47 @@ def split_by_episode(self) -> List["SampleBatch"]:
...
+        else:
+            raise KeyError(f"{self} does not have `eps_id` or `dones`!")
 
+        assert sum(s.count for s in slices) == self.count, (slices, self.count)
```
Comment (kouroshHakha) [inline rllib/policy/sample_batch.py:390]: Please add descriptive error messages here.

File: rllib/policy/tests/test_sample_batch.py [hunk 7]
```diff
@@ -263,6 +263,22 @@ def test_slicing(self):
...
 
+    def test_split_by_episode(self):
+        s = SampleBatch(
+            {
+                "a": np.array([0, 1, 2, 3, 4, 5]),
+                "eps_id": np.array([0, 0, 0, 0, 1, 1]),
+                "dones": np.array([0, 0, 0, 1, 0, 1]),
+            }
+        )
+        eps_split = [b["a"] for b in s.split_by_episode()]
```
Comment (kouroshHakha) [inline rllib/policy/tests/test_sample_batch.py:274]: Can you leave inline comments what is being tested in each section? 
Also can you modify your test logic to do the following?
1. check if your split_by_episode when episode exists returns the correct splits (hard-code the ground truth)
2. check if your split_by_done also returns the same thing (right now you check if it matches what ever split_by_eps_id returns. you don't check its absolute correctness.)
3. check if it properly raises error if neither done or eps_id exist
4. create a sample_batch where done is not properly set (all false or just false for the last portion), test if the behavior is expected.

**Review by kouroshHakha — DISMISSED** [review]
Thanks, @Rohan138 Please make the requested changes. O.w. looks good.

**Review by kouroshHakha — APPROVED** [review]

**Pull request merged**